### PR TITLE
Convert Existing UI to React

### PR DESCRIPTION
**Pull request by Vadifire**

### CODE REVIEW

File: src/js/components/TimelineUI.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		return (
+		    <div id="timeline-div">
+				<button id="update-timeline" onClick={() => this.fetchTweets()}>Update Home Timeline</button>
+				{displayedElem}
+			</div>
+		);
+	}
+
+	fetchTweets() {
+		fetch("http://localhost:8080/api/1.0/twitter/timeline").then(response => { // Attempt to fetch tweets
```
Comment (kev-chan) [inline src/js/components/TimelineUI.js]: why is this located in here?

File: src/js/twitter-api.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+			tweets.map(tweet => { // Populate default user to avoid reference error
+			if (!tweet.user) {
+				tweet.user = {
+					name: "Unknown Handle"
+				}
+			}
+			return tweet;
+		});
+			return {
+				tweets,
```
Comment (briansheen) [inline src/js/twitter-api.js]: what does tweets look like if there is no tweet.user?

File: src/js/twitter-api.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	}).then(tweets => { // Got JSON
+		if (tweets.length > 0) {
+			tweets.map(tweet => { // Populate default user to avoid reference error
+			if (!tweet.user) {
+				tweet.user = {
+					name: "Unknown Handle"
+				}
+			}
+			return tweet;
+		});
```
Comment (briansheen) [inline src/js/twitter-api.js]: formatting

File: src/js/components/TimelineUI.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		    <div id="timeline-div">
+				<button id="update-timeline" onClick={this.updateTimeline.bind(this)}>Update Home Timeline</button>
+				{displayedElem}
+			</div>
+		);
+	}
+
+	updateTimeline() {
+		fetchHomeTimeline().then(response => {
+			this.setState(response);
```
Comment (briansheen) [inline src/js/components/TimelineUI.js]: what is response? what if response has more k:v pairs than you expected?

File: src/js/components/TimelineUI.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+	render() {
+		let displayedElem; // Either display error or tweets
+		if (this.state.error) {
+			displayedElem = <div id="error-div">{this.state.error}</div>
+		}
+		else if (this.state.tweets) {
+			displayedElem = 
+				<div id="tweets">{
+					this.state.tweets.map((tweet, index) =>
```
Comment (briansheen) [inline src/js/components/TimelineUI.js]: do you need index here?

this is kind of hard to read, are you declaring another variable here?

File: src/js/twitter-api.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * Returns a promise with:
+ * {
+ *   tweets: [...]
+ *   error: String
+ * } 
+ * Invariant: Exactly one of tweets or error will be null
+ */
+export function fetchHomeTimeline() {
+	return fetch("http://localhost:8080/api/1.0/twitter/timeline").then(response => { // Attempt to fetch tweets
+		if (response.ok === true) {
```
Comment (briansheen) [inline src/js/twitter-api.js]: is there a better way to check if response is ok?

File: src/js/components/TimelineUI.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	componentDidMount() {
+		this.updateTimeline();
+	}
+
+	render() {
+		let displayedElem; // Either display error or tweets
+		if (this.state.error) {
+			displayedElem = <div id="error-div">{this.state.error}</div>
+		}
+		else if (this.state.tweets) {
```
Comment (briansheen) [inline src/js/components/TimelineUI.js]: do you need an else here?

File: src/js/components/TimelineUI.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		return (
+		    <div id="timeline-div">
+				<button id="update-timeline" onClick={()=>this.updateTimeline()}>Update Home Timeline</button>
+				{displayedElem}
+			</div>
+		);
+	}
+
+	updateTimeline() {
+		fetchHomeTimeline().then(response => {
```
Comment (kev-chan) [inline src/js/components/TimelineUI.js]: use proper promise syntax for handling errors

File: dist/index.html [hunk 9]
```diff
@@ -1,17 +1,9 @@
...
-		<div id="title">
-			Lab for Cedric
-		</div>
-		<div id="timeline-div">
-			<button id="update-timeline">Update Home Timeline</button>
-			<div id="error-div"></div>
-			<div id="tweets" class="hidden"></div>
-		</div>
+		<div id="root"></div>
+		<script src="js/bundle.js"></script>
```
Comment (kev-chan) [inline dist/index.html]: why was this moved?

File: src/js/twitter-api.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Fetches tweets from home timeline
+ *
+ * Returns a promise wrapping tweets
+ */
+export function fetchHomeTimeline() {
+	return fetch("http://localhost:8080/api/1.0/twitter/timeline").then(response => { // Attempt to fetch tweets
+		if (!response.ok) {
```
Comment (kev-chan) [inline src/js/twitter-api.js]: what is this code supposed to be doing?

File: src/js/components/TimelineUI.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
+import React, { Component } from "react";
```
Comment (kev-chan) [inline src/js/components/TimelineUI.js]: please ensure Brian's comments are addressed

File: src/js/components/TimelineUI.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+						<div key={tweet.url} className="row">
+							<TweetBlock tweet={tweet}/>
+						</div>
+					)
+				}
+			</div>;
+		} 
+		return (
+		    <div id="timeline-div">
+				<button id="update-timeline" onClick={this.updateTimeline.bind(this)}>Update Home Timeline</button>
```
[the remark this answers is not in the document]
Comment (Vadifire) [reply src/js/components/TimelineUI.js]: I want to bind the function here to ensure I can use 'this' in updateTimeline. Is there a better way to solve this problem?

File: src/js/components/TimelineUI.js (hunk 4 above)
Comment (Vadifire) [reply src/js/components/TimelineUI.js]: The response here is an array of tweet objects. Is the concern that I could be storing data to state that I'm not using if the tweet objects have extra k:v pairs?

File: src/js/components/TimelineUI.js (hunk 5 above)
Comment (Vadifire) [reply src/js/components/TimelineUI.js]: I'm creating a React component for the tweets and assigning it to displayedElem here. I agree it's hard to read. I changed this to the code snippet below; is this easier to read?
`
const tweets = this.state.tweets.map(tweet =>
	<div key={tweet.url} className="row">
		<TweetBlock tweet={tweet}/>
	</div>
);
displayedElem = <div id="tweets">{tweets}</div>;
`

Comment (Vadifire) [reply src/js/components/TimelineUI.js]: I'm creating a React component for the tweets and assigning it to displayedElem here. I agree it's hard to read. Does the updated code look cleaner now?

File: src/js/components/TimelineUI.js (hunk 4 above)
Comment (briansheen) [reply src/js/components/TimelineUI.js]: yep, do you want to be setting state in your timelineUI component for things it may potentially not care about?

File: src/js/components/TimelineUI.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (briansheen) [reply src/js/components/TimelineUI.js]: yeah the binding is good.  but where and why do you want to bind?

File: src/js/components/TimelineUI.js (hunk 5 above)
Comment (briansheen) [reply src/js/components/TimelineUI.js]: 👍

File: src/js/twitter-api.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+	}).then(tweets => { // Got JSON
+		tweets.map(tweet => { // Populate default user to avoid reference error
+			if (!tweet.user) {
+				tweet.user = {
+					name: "Unknown User"
+				}
+			}
+			return tweet;
+		});
+		return tweets;
```
Comment (briansheen) [inline src/js/twitter-api.js]: if you added some default users, what is tweets here?

File: src/js/components/TimelineUI.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Vadifire) [reply src/js/components/TimelineUI.js]: Ah, binding should be in the constructor to avoid creating a different callback during each render(). Thanks.

File: src/js/twitter-api.js (hunk 13 above)
Comment (Vadifire) [reply src/js/twitter-api.js]: Tweets is still an array of tweet objects, but the "default" users would only have a name. I added empty Strings for displayHandle and profileImageUrl to help avoid errors down the road.

Comment (briansheen) [reply src/js/twitter-api.js]: did you test to make sure this happens?

Comment (Vadifire) [reply src/js/twitter-api.js]: Yes, but maybe I'm misunderstanding what you mean by "default user". I tested adding tweets with null/undefined users and the default values populated fine. Are you referring to a different case?

File: dist/index.html (hunk 9 above)
Comment (Vadifire) [reply dist/index.html]: Are you talking about the script tag? I moved it to the bottom of the body to ensure the root element is rendered to the DOM before attempting to add any elements to it. Was the previous method better?

File: src/js/components/TimelineUI.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+				<button id="update-timeline" onClick={this.updateTimeline}>Update Home Timeline</button>
+				{displayedElem}
+			</div>
+		);
+	}
+
+	// Property 'fetchTimeline' must point to function that returns Promise with tweets
+	updateTimeline() {
+		this.props.fetchTimeline().then(tweets => {
+			if (tweets.length > 0) {
```
Comment (kev-chan) [inline src/js/components/TimelineUI.js]: shorthand this

File: src/js/components/Application.js [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+import "../../css/components/Application.scss";
+
+// Contains Title, Hello Message, and Timeline UI
+export default class Application extends React.Component {
+	render() {
+		return (
+			<div>
+				<div id="title">Lab for Cedric</div>
+				<HelloMessage message="hello react!" />
+				<TimelineUI fetchTimeline={fetchHomeTimeline}/>
```
Comment (kev-chan) [inline src/js/components/Application.js]: why is fetchTimeline being passed in?

File: src/js/components/TimelineUI.js (hunk 14 above)
Comment (kev-chan) [reply src/js/components/TimelineUI.js]: this still has not been handled...
